mobile_verifier/src/rewarder.rs

```diff
@@ -0,0 +1,226 @@
+    ingest,
+    reward_shares::{PocShares, TransferRewards},
+    speedtests::SpeedtestAverages,
+};
+use anyhow::bail;
+use chrono::{DateTime, Duration, TimeZone, Utc};
+use db_store::meta;
+use file_store::{file_sink::FileSinkClient, traits::TimestampEncode, FileStore};
+use helium_proto::RewardManifest;
+use price::PriceTracker;
+use rust_decimal::{prelude::ToPrimitive, Decimal};
+use rust_decimal_macros::dec;
+use sqlx::{PgExecutor, Pool, Postgres};
+use std::ops::Range;
+use tokio::time::sleep;
+
+pub struct Rewarder {
+    pool: Pool<Postgres>,
+    reward_period_duration: Duration,
+    reward_offset: Duration,
+    mobile_rewards: FileSinkClient,
+    reward_manifests: FileSinkClient,
+    price_tracker: PriceTracker,
+    data_transfer_ingest: FileStore,
+}
+
+impl Rewarder {
+    pub fn new(
+        pool: Pool<Postgres>,
+        reward_period_duration: Duration,
+        reward_offset: Duration,
+        mobile_rewards: FileSinkClient,
+        reward_manifests: FileSinkClient,
+        price_tracker: PriceTracker,
+        data_transfer_ingest: FileStore,
+    ) -> Self {
+        Self {
+            pool,
+            reward_period_duration,
+            reward_offset,
+            mobile_rewards,
+            reward_manifests,
+            price_tracker,
+            data_transfer_ingest,
+        }
+    }
+
+    pub async fn run(self, shutdown: triggered::Listener) -> anyhow::Result<()> {
+        loop {
+            let last_rewarded_end_time = self.last_rewarded_end_time().await?;
+            let next_rewarded_end_time = self.next_rewarded_end_time().await?;
+            let now = Utc::now();
+            let next_reward = if now > (next_rewarded_end_time + self.reward_offset) {
+                Duration::zero().to_std()?
+            } else {
+                let next_reward =
+                    (next_rewarded_end_time - Utc::now() + self.reward_offset).to_std()?;
+                tracing::info!(
+                    "Next reward will be given in {}",
+                    humantime::format_duration(next_reward)
+                );
+                next_reward
+            };
+            tokio::select! {
+                _ = shutdown.clone() => break,
+                _ = sleep(next_reward) =>
+                    self.reward(&(last_rewarded_end_time..next_rewarded_end_time)).await?,
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn last_rewarded_end_time(&self) -> db_store::Result<DateTime<Utc>> {
+        Utc.timestamp_opt(meta::fetch(&self.pool, "last_rewarded_end_time").await?, 0)
+            .single()
+            .ok_or(db_store::Error::DecodeError)
+    }
+
+    async fn next_rewarded_end_time(&self) -> db_store::Result<DateTime<Utc>> {
+        Utc.timestamp_opt(meta::fetch(&self.pool, "next_rewarded_end_time").await?, 0)
+            .single()
+            .ok_or(db_store::Error::DecodeError)
+    }
+
+    async fn disable_complete_data_checks_until(&self) -> db_store::Result<DateTime<Utc>> {
```

Comment:
Shouldn't we default this if the key is not found in the meta table?

Reply:
I have two opinions about this. The first is that we should probably _not_ default it, as it's a very important setting. The second is that _if_ we want to default this, we might as well default it by inserting the key in a migration

Reply:
I can get behind setting a default via migration, but I think we have to have a default or catch the error where it does not exist.

Reply:
Added default in a new migration